File: app/src/main/java/com/mandarin/bcu/androidutil/medal/adapters/MedalListAdapter.java
package com.mandarin.bcu.androidutil.medal.adapters;

import android.app.Activity;
import android.app.Dialog;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.ArrayAdapter;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.mandarin.bcu.R;
import com.mandarin.bcu.androidutil.StaticStore;
import com.mandarin.bcu.androidutil.adapters.SingleClick;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class MedalListAdapter extends ArrayAdapter<String> {
    private final Activity activity;
    private int num;
    private int height;

    private float imgwh;

    public MedalListAdapter(Activity activity, int num, int width, float imgwh, String[] lines) {
        super(activity, R.layout.medal_layout, lines);
        this.activity = activity;
        this.num = num;
        this.imgwh = imgwh;

        height = width / num - StaticStore.dptopx(4f, activity);
    }

    private static class ViewHolder {
        LinearLayout layout;
        List<ImageView> icons;

        private ViewHolder(View view) {
            layout = view.findViewById(R.id.medallinear);
            icons = new ArrayList<>();
        }
    }

    @NonNull
    @Override
    public View getView(int position, View view, @NonNull ViewGroup group) {
        View row;
        ViewHolder holder;

        if (view == null) {
            LayoutInflater inflater = LayoutInflater.from(activity);
            row = inflater.inflate(R.layout.medal_layout, group, false);
            holder = new ViewHolder(row);
            row.setTag(holder);
        } else {
            row = view;
            holder = (ViewHolder) row.getTag();
        }

        int posit = num * position;

        holder.layout.removeAllViews();
        holder.icons.clear();

        for (int j = 0; j < num; j++) {

            if (posit + j < StaticStore.medals.size()) {
                final int jj = j;
                Bitmap b = StaticStore.getResizeb(StaticStore.medals.get(posit + j), activity, imgwh);

                ImageButton icon = new ImageButton(activity);
                icon.setBackground(activity.getDrawable(R.drawable.image_button_circular));
                LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT, 1.0f);
                params.height = height;
                params.setMarginStart(StaticStore.dptopx(2f, activity));
                params.setMarginEnd(StaticStore.dptopx(2f, activity));
                icon.setLayoutParams(params);

                icon.setImageBitmap(b);

                icon.setOnClickListener(new SingleClick() {
                    @Override
                    public void onSingleClick(View v) {
                        Dialog dialog = new Dialog(activity);

                        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
                        dialog.setContentView(R.layout.layout_medal_desc);
                        Objects.requireNonNull(dialog.getWindow()).setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));

                        View v1 = dialog.getWindow().getDecorView();
                        v1.setBackground(activity.getDrawable(R.drawable.dialog_box));

                        ImageView icon = dialog.findViewById(R.id.medalimg);
                        TextView name = dialog.findViewById(R.id.medalname);
                        TextView desc = dialog.findViewById(R.id.medaldesc);

                        icon.setImageBitmap(StaticStore.getResizeb(StaticStore.medals.get(posit + jj), activity, imgwh));
                        name.setText(StaticStore.MEDNAME.getCont(posit + jj));
                        desc.setText(StaticStore.MEDEXP.getCont(posit + jj));

                        WindowManager.LayoutParams lp = new WindowManager.LayoutParams();
                        lp.copyFrom(dialog.getWindow().getAttributes());
                        lp.width = WindowManager.LayoutParams.MATCH_PARENT;
                        lp.height = WindowManager.LayoutParams.WRAP_CONTENT;

                        dialog.getWindow().setAttributes(lp);

                        dialog.show();
                    }
                });

                holder.layout.addView(icon);
                holder.icons.add(icon);
            } else {
                Bitmap b = StaticStore.empty(activity, imgwh, imgwh);

                ImageView icon = new ImageView(activity);
                icon.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT, 1.0f));
                icon.setImageBitmap(b);

                holder.layout.addView(icon);
                holder.icons.add(icon);
            }
        }

        return row;
    }
}

File: app/src/main/java/com/mandarin/bcu/androidutil/medal/adapters/MedalListAdapter.kt
package com.mandarin.bcu.androidutil.medal.adapters

import android.app.Activity
import android.app.Dialog
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.view.*
import android.widget.*
import androidx.core.content.ContextCompat
import com.mandarin.bcu.R
import com.mandarin.bcu.androidutil.StaticStore
import com.mandarin.bcu.androidutil.supports.SingleClick
import kotlin.collections.ArrayList

class MedalListAdapter(private val activity: Activity, private val num: Int, width: Int, private val imgwh: Float, lines: Array<String?>, private val order: ArrayList<Int>) : ArrayAdapter<String?>(activity, R.layout.medal_layout, lines) {
    private val height: Int = width / num - StaticStore.dptopx(4f, activity)

    private class ViewHolder constructor(view: View) {
        var layout: LinearLayout = view.findViewById(R.id.medallinear)
        var icons: MutableList<ImageView> = ArrayList()

    }

    override fun getView(position: Int, view: View?, group: ViewGroup): View {
        val holder: ViewHolder
        val row: View

        if (view == null) {
            val inf = LayoutInflater.from(context)
            row = inf.inflate(R.layout.medal_layout, group, false)
            holder = ViewHolder(row)
            row.tag = holder
        } else {
            row = view
            holder = row.tag as ViewHolder
        }


        val posit = num * position

        holder.layout.removeAllViews()
        holder.icons.clear()

        for (j in 0 until num) {
            if (posit + j < StaticStore.medals.size) {
                val b = StaticStore.getResizeb(StaticStore.medals[posit + j], activity, imgwh)
                val icon = ImageButton(activity)
                icon.background = ContextCompat.getDrawable(activity, R.drawable.image_button_circular)
                val params = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT, 1.0f)
                params.height = height
                params.marginStart = StaticStore.dptopx(2f, activity)
                params.marginEnd = StaticStore.dptopx(2f, activity)
                icon.layoutParams = params
                icon.setImageBitmap(b)
                icon.setOnClickListener(object : SingleClick() {
                    override fun onSingleClick(v: View?) {
                        val dialog = Dialog(activity)
                        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE)
                        dialog.setContentView(R.layout.layout_medal_desc)
                        dialog.window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
                                ?: return
                        val v1 = dialog.window?.decorView ?: return
                        v1.background = ContextCompat.getDrawable(activity, R.drawable.dialog_box)
                        val icon1 = dialog.findViewById<ImageView>(R.id.medalimg)
                        val name = dialog.findViewById<TextView>(R.id.medalname)
                        val desc = dialog.findViewById<TextView>(R.id.medaldesc)
                        icon1.setImageBitmap(StaticStore.getResizeb(StaticStore.medals[posit + j], activity, imgwh))
                        name.text = StaticStore.MEDNAME.getCont(order[posit + j])
                        desc.text = StaticStore.MEDEXP.getCont(order[posit + j])
                        val lp = WindowManager.LayoutParams()
                        lp.copyFrom(dialog.window?.attributes)
                        lp.width = WindowManager.LayoutParams.MATCH_PARENT
                        lp.height = WindowManager.LayoutParams.WRAP_CONTENT
                        dialog.window?.attributes = lp
                        dialog.show()
                    }
                })
                holder.layout.addView(icon)
                holder.icons.add(icon)
            } else {
                val b = StaticStore.empty(activity, imgwh, imgwh)
                val icon = ImageView(activity)
                icon.layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT, 1.0f)
                icon.setImageBitmap(b)
                holder.layout.addView(icon)
                holder.icons.add(icon)
            }
        }
        return row
    }

}